ale/teacher/exploitation/tag_flip_historical.py

```diff
@@ -0,0 +1,144 @@
+from ale.trainer.base_trainer import Predictor, PredictionResult
+from ale.trainer.prediction_result import TokenConfidence, LabelConfidence
+
+class HistoricalDoc:
+    def __init__(self, idx: int, prediction_result: PredictionResult) -> None:
+        """
+        Helper class used to save predictions of last cycle for one document.
+        """
+        self.idx: int = idx
+        labels: List[str] = [token_confidence.get_predicted_label() for token_confidence in prediction_result.ner_confidences_token]
+        self.labels: List[str] = labels
+
+    def update_and_compare_labels(self, predictions: PredictionResult) -> List[int]:
+        """
+        Compares old and new labels and returns a list indicating which label predictions have changed (per token).
+
+        Args:
+            - labels (List[str]): a list of new labels for each token
+
+        Returns:
+            - difference (List[int]): a list which indicates if the token-predictions has changed (1) or not (0)
+        """
+        difference: List[int] = []
+        labels: List[str] = [token_confidence.get_predicted_label() for token_confidence in predictions.ner_confidences_token]
+        assert len(labels) == len(self.labels)
+        for i,label in enumerate(labels):
+            if label == self.labels[i]:
+                difference.append(0) # no tag flip
+            else:
+                difference.append(1) # tag flip
+        self.labels = labels
+        return difference
+
+    def get_labels(self) -> List[str]:
+        return self.labels
+
+class HistoricalSequence:
+    def __init__(self, docs_with_predictions: Dict[int,PredictionResult]) -> None:
+        """
+        Helper class used to wrap predictions of last cycle.
+
+        Args:
+            - docs_with_predictions (Dict[int,List[str]]): Dict containing the ids of the documents and their predicted labels as a list
+        """
+        self.docs: List[HistoricalDoc] = []
+        for doc in docs_with_predictions:
+            historical_doc: HistoricalDoc = HistoricalDoc(doc,docs_with_predictions[doc])
+            self.docs.append(historical_doc)
+    
+    def update_and_compare_historical_sequences(self, docs_with_predictions: Dict[int,PredictionResult]) -> Dict[int,List[int]]:
+        differences: Dict[int,List[int]] = {}
+        for doc in docs_with_predictions:
+            historical_doc: Optional[HistoricalDoc] = self.get_historical_doc_by_id(doc)
+            if historical_doc: # history exists for document
+                difference: List[int] = historical_doc.update_and_compare_labels(docs_with_predictions[doc])
+                differences[doc] = difference
+            else:
+                historical_doc: HistoricalDoc = HistoricalDoc(doc,docs_with_predictions[doc])
+                difference: List[int] = [0 for item in historical_doc.get_labels()] # doc never predicted before, so no difference
+                differences[doc] = difference
+        return differences
```

Comment:
The strategy works only on n and n-1? Not n-2 or n-x past predictions? Updates override the old ones

Reply:
Yes, this is how the authors implemented it if I get that right.

Reply:
You are right. It could be something we could test in the end, if this type of strategy performs well.